File: gen_coverage/gcov.py
import os
import shutil
import glob
import json
import hashlib
import subprocess

from collections import defaultdict
from pathlib import Path

from . import GCOV_PREFIX_BASE
from .utils import combine_reports
from .fastcov import distillSource

def symlink_gcno_files(build_dir, prefix_dir, verbose=False):
    if not prefix_dir or prefix_dir == "/":
        if verbose:
            print("Empty prefix, early return")
        return

    build_dir = Path(build_dir).absolute()
    prefix_dir = Path(prefix_dir).absolute()
    gcno_files = glob.glob(os.path.join(build_dir, '**', '*.gcno'), recursive=True)

    for f in gcno_files:
        # Create target directory if it doesn't exist
        os.makedirs(str(prefix_dir) + os.path.dirname(f), exist_ok=True)

        target_file = str(prefix_dir) + f
        # Create symlink if it doesn't already exist
        if not os.path.exists(target_file):
            os.symlink(f, target_file)
            if verbose:
                print(f"Created symlink: {target_file} -> {f}")
        else:
            if verbose:
                print(f"Symlink already exists: {target_file}")


def gcov_init(bench_dir):
    shutil.rmtree(GCOV_PREFIX_BASE, ignore_errors=True)
    os.makedirs(GCOV_PREFIX_BASE, exist_ok=True)

    # Also make sure to create the per file dirs
    all_files = glob.glob(os.path.join(bench_dir, "**/*.smt2"), recursive=True)
    for file in all_files:
        dir_path = get_prefix(file)
        os.makedirs(dir_path, exist_ok=True)
        

def gcov_cleanup():
    shutil.rmtree(GCOV_PREFIX_BASE, ignore_errors=True)

def get_file_uid(file):
    hash = hashlib.sha256(str(file).encode()).hexdigest()
    h_readable = ''.join(Path(file).parts[-2:])
    if h_readable.endswith(".smt2"): 
        h_readable = h_readable[:-5]
    if len(h_readable) > 20:
        h_readable = h_readable[-20:]

    return f"{hash}-{h_readable}"

def get_prefix(file):
    return os.path.join(GCOV_PREFIX_BASE, get_file_uid(file))

def get_gcov_env(file):
    env = os.environ.copy()
    env["GCOV_PREFIX"] = get_prefix(file)

    return env

def get_prefix_files(prefix=GCOV_PREFIX_BASE):
    path_wildcard = os.path.join(prefix, "**/*.gcda")
    return glob.glob(path_wildcard, recursive=True)

def process_prefix(prefix, files, verbose=False):
    files_report = { "sources": {} }
    # print(f"process_prefix: len(files) {len(files)}")
    for gcda_file in files:
        env = os.environ.copy()
        env["GCOV_PREFIX"] = prefix
        if verbose:
            print("Gcov GCDA File:" + str(gcda_file))
        result = subprocess.run(['gcov', '--json', '--stdout', gcda_file], env=env, check=False, capture_output=True, text=True)
        next_report = {"sources": {}}

        store_noisy_branches = False
        if verbose:
            print("Gcov Exit Code: " + str(result.returncode))
            print("Gcov Errors: " + str(result.stderr or None))

        source = json.loads(result.stdout)
        for f in source["files"]:
            f_path = gcda_file[:-5] if gcda_file.endswith(".gcda") else gcda_file
            f_path = f_path[len(prefix):] if f_path.startswith(prefix) else f_path
            f["file_abs"] = f_path
            distillSource(f, next_report["sources"], "", store_noisy_branches)

        # Merge files together using our special "per-test" counter
        # print(f"process_prefix: combining reports (exec_one=True): \n curr report:" + str(files_report)[0:100] + '\n------------------------------------------\n next report:' + str(next_report)[0:100])
        combine_reports(files_report, next_report, exec_one=True)

    # print(f"process_prefix: files_report (result): " + str(files_report)[0:200])
    return files_report

File: rust/src/gcov.rs
use glob::glob;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::io::{self};
use std::os::unix::fs::symlink;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::process::Output;

use crate::args::ARGS;
use crate::utils::combine_reports;

// External constants (Replace with the actual constant value)
const GCOV_PREFIX_BASE: &str = "/path/to/gcov_prefix_base";

// Initialize gcov directories and cleanup old data
pub fn init() -> io::Result<()> {
    fs::create_dir_all(GCOV_PREFIX_BASE)?;

    let pattern = format!(
        "{}/**/*.smt2",
        ARGS.benchmark_dir
            .canonicalize()
            .unwrap()
            .display()
            .to_string()
    );
    for entry in glob(&pattern).expect("Failed to read glob pattern") {
        if let Ok(file) = entry {
            let dir_path = get_prefix(&file);
            fs::create_dir_all(&dir_path)?;
        }
    }

    Ok(())
}

// Cleanup gcov data
pub fn cleanup() -> io::Result<()> {
    fs::remove_dir_all(GCOV_PREFIX_BASE)?;
    Ok(())
}

// Symlink `.gcno` files from build directory to prefix directory
pub fn symlink_gcno_files(prefix_dir: &Path) -> io::Result<()> {
    if prefix_dir == Path::new("/") {
        if ARGS.verbose {
            println!("Empty prefix, early return");
        }
        return Ok(());
    }

    let build_dir = fs::canonicalize(&ARGS.build_dir)?;
    let prefix_dir = fs::canonicalize(prefix_dir)?;

    let pattern = format!("{}/**/*.gcno", build_dir.to_string_lossy());
    for entry in glob(&pattern).expect("Failed to read glob pattern") {
        if let Ok(gcno_file) = entry {
            let target_dir = prefix_dir.join(gcno_file.parent().unwrap());
            fs::create_dir_all(&target_dir)?;

            let target_file = prefix_dir.join(&gcno_file);
            if !target_file.exists() {
                symlink(&gcno_file, &target_file)?;
                if ARGS.verbose {
                    println!(
                        "Created symlink: {} -> {}",
                        target_file.display(),
                        gcno_file.display()
                    );
                }
            } else if ARGS.verbose {
                println!("Symlink already exists: {}", target_file.display());
            }
        }
    }
    Ok(())
}

// Get a unique file identifier using SHA-256 and the file path
pub fn get_file_uid(file: &Path) -> String {
    let mut hasher = Sha256::new();
    hasher.update(file.to_string_lossy().as_bytes());
    let hash = format!("{:x}", hasher.finalize());

    let mut h_readable: String = file
        .components()
        .rev()
        .take(2)
        .map(|comp| comp.as_os_str().to_string_lossy().to_string())
        .collect();

    if h_readable.ends_with(".smt2") {
        h_readable.truncate(h_readable.len() - 5);
    }
    if h_readable.len() > 20 {
        h_readable.truncate(20);
    }

    format!("{}-{}", hash, h_readable)
}

// Get the GCOV_PREFIX for a file
pub fn get_prefix(file: &Path) -> PathBuf {
    Path::new(GCOV_PREFIX_BASE).join(get_file_uid(file))
}

// Prepare environment variables for running gcov
pub fn get_gcov_env(file: &Path) -> HashMap<String, String> {
    let mut env = env::vars().collect::<HashMap<_, _>>();
    env.insert(
        "GCOV_PREFIX".to_string(),
        get_prefix(file).to_string_lossy().to_string(),
    );
    env
}

// Get all gcda files from the given prefix
pub fn get_prefix_files(prefix: &Path) -> Result<Vec<PathBuf>, Box<dyn std::error::Error>> {
    let pattern = format!("{}/**/*.gcda", prefix.canonicalize()?.display());
    let files = glob(&pattern)?
        .filter_map(Result::ok)
        .collect::<Vec<PathBuf>>();

    Ok(files)
}

// Process a prefix by running gcov on the files and combining the results
pub fn process_prefix(prefix: &Path, files: &Vec<PathBuf>) -> io::Result<Value> {
    let mut files_report = HashMap::new();
    files_report.insert("sources".to_string(), serde_json::json!({}));
    let mut files_report_value = Value::Object(files_report.into_iter().collect());

    for gcda_file in files {
        let mut env = env::vars().collect::<HashMap<_, _>>();
        env.insert(
            "GCOV_PREFIX".to_string(),
            prefix.canonicalize()?.display().to_string(),
        );

        if ARGS.verbose {
            println!("Gcov GCDA File: {}", gcda_file.display());
        }

        let output = run_gcov(gcda_file, &env)?;

        if ARGS.verbose {
            println!("Gcov Exit Code: {}", output.status);
            println!("Gcov Errors: {}", String::from_utf8_lossy(&output.stderr));
        }

        let source: Value = serde_json::from_slice(&output.stdout)?;
        let next_report = process_gcov_output(&source)?;
        // FIXME: This should call distillSource

        let next_report_value = Value::Object(next_report.into_iter().collect());

        combine_reports(&mut files_report_value, &next_report_value, true);
    }

    Ok(files_report_value)
}

// Function to run the gcov command on a file
fn run_gcov(gcda_file: &Path, env: &HashMap<String, String>) -> io::Result<Output> {
    Command::new("gcov")
        .args(&["--json", "--stdout", gcda_file.to_str().unwrap()])
        .envs(env)
        .output()
}

// Process the gcov output to distill sources (dummy function)
fn process_gcov_output(source: &Value) -> io::Result<HashMap<String, Value>> {
    // Placeholder for processing gcov output
    Ok(HashMap::new())
}
